src/buffer/out/LineRendition.hpp

```diff
@@ -0,0 +1,34 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+
+Module Name:
+- LineRendition.hpp
+
+Abstract:
+- Enumerated type for the VT line rendition attribute. This determines the
+  width and height scaling with which each line is rendered.
+
+--*/
+
+#pragma once
+
+enum class LineRendition
+{
+    SingleWidth,
+    DoubleWidth,
+    DoubleHeightTop,
+    DoubleHeightBottom
+};
+
+constexpr SMALL_RECT ScreenToBufferLine(const SMALL_RECT& line, const LineRendition lineRendition)
+{
+    const auto scale = lineRendition == LineRendition::SingleWidth ? 0 : 1;

```

Comment:
Yeah, there's probably an element of premature optimization here. I was less concerned about speed though - I honestly wouldn't know if an imul or idiv is significantly worse than a shift on a modern processor. But I was horrified by the amount of code I was introducing in my initial implementation, and the shifts were definitely an improvement in that respect.

The current version is actually a bit of a compromise - the optimizer doesn't handle inlining of these functions very well - but I was also trying to keep some of the messy details hidden in the helper functions, and not bloat the rest of the source too much.

There was also another reason I preferred the shifts - at least for division. If the values are negative, then a div doesn't round in the right direction, whereas a shift does. I don't _think_ we need to support negative numbers in the current usage, but it felt safer knowing that these functions would do the right thing if that ever did occur.

I agree this could definitely do with some commenting though.